# Data Science Challenge - Mercado Libre

## 3. Failure predict

## 📥 Data Import and Initial Exploration

This section focuses on importing required libraries, loading the dataset, and performing a basic exploratory analysis.

---

In [4]:
# --- Imports ---
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, confusion_matrix
from imblearn.over_sampling import SMOTE
from imblearn.ensemble import BalancedRandomForestClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from xgboost import XGBClassifier
from sklearn.ensemble import HistGradientBoostingClassifier
from lightgbm import LGBMClassifier
# Load the dataset
file_path = 'full_devices.csv'
df = pd.read_csv(file_path, encoding='latin1')

# Display basic info and first few rows
df.info()
# Preview unique devices and columns
unique_devices = df['device'].nunique()
print(f"Number of unique devices: {unique_devices}")
# Display covered period
covered_period = df['date'].min(), df['date'].max()
print(f"Covered period: {covered_period}")
# Evaluate if all devices in the database had at least one failure
devices_with_failure = df.groupby('device')['failure'].max()
all_failed = devices_with_failure.all()
print(f"Did all devices have at least one failure? {all_failed}")
print(f"Number of devices that never failed: {(devices_with_failure == 0).sum()}")
# Number of devices that had at least one failure
num_failed_devices = (devices_with_failure > 0).sum()
print(f"Number of devices that had at least one failure: {num_failed_devices}")
df.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124494 entries, 0 to 124493
Data columns (total 12 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   date        124494 non-null  object
 1   device      124494 non-null  object
 2   failure     124494 non-null  int64 
 3   attribute1  124494 non-null  int64 
 4   attribute2  124494 non-null  int64 
 5   attribute3  124494 non-null  int64 
 6   attribute4  124494 non-null  int64 
 7   attribute5  124494 non-null  int64 
 8   attribute6  124494 non-null  int64 
 9   attribute7  124494 non-null  int64 
 10  attribute8  124494 non-null  int64 
 11  attribute9  124494 non-null  int64 
dtypes: int64(10), object(2)
memory usage: 11.4+ MB
Number of unique devices: 1169
Covered period: ('2015-01-01', '2015-11-02')
Did all devices have at least one failure? False
Number of devices that never failed: 1063
Number of devices that had at least one failure: 106


,date,device,failure,attribute1,attribute2,attribute3,attribute4,attribute5,attribute6,attribute7,attribute8,attribute9
0,2015-01-01,S1F01085,0,215630672,56,0,52,6,407438,0,0,7
1,2015-01-01,S1F0166B,0,61370680,0,3,0,6,403174,0,0,0
2,2015-01-01,S1F01E6Y,0,173295968,0,0,0,12,237394,0,0,0
3,2015-01-01,S1F01JE0,0,79694024,0,0,0,6,410186,0,0,0
4,2015-01-01,S1F01R2B,0,135970480,0,0,0,15,313173,0,0,3


## ⚙️ Data Preparation and Modeling Pipeline

This section handles the full preparation process before training the models: creating the prediction target, generating features, stratified splitting, balancing the classes, and defining/testing machine learning models.

---

In [20]:
# --- Convert 'date' column to datetime format and sort by device and date ---
df['date'] = pd.to_datetime(df['date'])
df = df.sort_values(['device', 'date']).reset_index(drop=True)

# --- Create the target column: indicates if the device will fail on the next day ---
# This shifts the 'failure' column one day backward per device group
df['target'] = df.groupby('device')['failure'].shift(-1).fillna(0)

# --- Generate Rolling Statistics (mean and std) for each attribute per device ---
rolling_window = 5  # Size of the rolling window in days
attr_cols = [f'attribute{i}' for i in range(1, 10)]

# Compute rolling mean and standard deviation
for col in attr_cols:
    df[f'{col}_roll_mean'] = df.groupby('device')[col].transform(
        lambda x: x.rolling(window=rolling_window, min_periods=1).mean()
    )
    df[f'{col}_roll_std'] = df.groupby('device')[col].transform(
        lambda x: x.rolling(window=rolling_window, min_periods=1).std().fillna(0)
    )

# --- Stratified split: divide devices into training and test sets, maintaining class balance ---
device_failures = df.groupby('device')['failure'].max()
failed_devices = device_failures[device_failures == 1].index
healthy_devices = device_failures[device_failures == 0].index

# Use 20% of each group for the test set
train_failed, test_failed = train_test_split(failed_devices, test_size=0.2, random_state=42)
train_healthy, test_healthy = train_test_split(healthy_devices, test_size=0.2, random_state=42)

# Combine failed and healthy devices for training and test
train_devices = list(train_failed) + list(train_healthy)
test_devices = list(test_failed) + list(test_healthy)

# Filter rows by selected devices
train_df = df[df['device'].isin(train_devices)].copy()
test_df = df[df['device'].isin(test_devices)].copy()

# --- Prepare Feature Matrix ---
# Include raw attributes and their rolling mean and std
feature_cols = attr_cols + [f'{col}_roll_mean' for col in attr_cols] + [f'{col}_roll_std' for col in attr_cols]

X_train = train_df[feature_cols]
y_train = train_df['target']
X_test = test_df[feature_cols]
y_test = test_df['target']

# --- Scale features (important for many models, especially Logistic Regression and SVMs) ---
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# --- Apply SMOTE to balance the classes in the training data ---
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X_train_scaled, y_train)

# --- Define candidate models to train and evaluate ---
models = {
    "Random Forest": RandomForestClassifier(random_state=42),
    "Balanced RF": BalancedRandomForestClassifier(random_state=42),
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "XGBoost": XGBClassifier(use_label_encoder=False, eval_metric='logloss'),
    "HistGradientBoosting": HistGradientBoostingClassifier(),
    "LightGBM": LGBMClassifier()
}

def evaluate_model(name, model, X_test, y_test):
    """
    Train and evaluate a given model, printing its performance.

    Parameters:
        name (str): The name of the model (for display).
        model (sklearn-like estimator): The trained model to evaluate.
        X_test (np.array): Test features.
        y_test (np.array or Series): Ground truth test labels.

    Prints:
        Confusion matrix and classification report including precision, recall, and F1-score.
    """
    y_pred = model.predict(X_test)
    print(f"\n=== {name} ===")
    print(confusion_matrix(y_test, y_pred))
    print(classification_report(y_test, y_pred, digits=4))

# --- Train each model and evaluate on the same test set ---
for name, model in models.items():
    model.fit(X_resampled, y_resampled)
    evaluate_model(name, model, X_test_scaled, y_test)



=== Random Forest ===
[[25337   118]
 [   22     0]]
              precision    recall  f1-score   support

         0.0     0.9991    0.9954    0.9972     25455
         1.0     0.0000    0.0000    0.0000        22

    accuracy                         0.9945     25477
   macro avg     0.4996    0.4977    0.4986     25477
weighted avg     0.9983    0.9945    0.9964     25477


=== Balanced RF ===
[[25348   107]
 [   21     1]]
              precision    recall  f1-score   support

         0.0     0.9992    0.9958    0.9975     25455
         1.0     0.0093    0.0455    0.0154        22

    accuracy                         0.9950     25477
   macro avg     0.5042    0.5206    0.5064     25477
weighted avg     0.9983    0.9950    0.9966     25477


=== Logistic Regression ===
[[24054  1401]
 [    7    15]]
              precision    recall  f1-score   support

         0.0     0.9997    0.9450    0.9716     25455
         1.0     0.0106    0.6818    0.0209        22

    accuracy    

c:\Users\Robson\AppData\Local\Programs\Python\Python313\Lib\site-packages\xgboost\training.py:183: UserWarning: [19:08:43] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)



=== XGBoost ===
[[25384    71]
 [   21     1]]
              precision    recall  f1-score   support

         0.0     0.9992    0.9972    0.9982     25455
         1.0     0.0139    0.0455    0.0213        22

    accuracy                         0.9964     25477
   macro avg     0.5065    0.5213    0.5097     25477
weighted avg     0.9983    0.9964    0.9973     25477


=== HistGradientBoosting ===
[[25313   142]
 [   18     4]]
              precision    recall  f1-score   support

         0.0     0.9993    0.9944    0.9968     25455
         1.0     0.0274    0.1818    0.0476        22

    accuracy                         0.9937     25477
   macro avg     0.5133    0.5881    0.5222     25477
weighted avg     0.9985    0.9937    0.9960     25477

[LightGBM] [Info] Number of positive: 98933, number of negative: 98933
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.008178 seconds.
You can set `force_row_wise=true` to remove the overhead.
And 

c:\Users\Robson\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


## 🔍 Model Comparison for Device Failure Prediction

This evaluation compares six machine learning models for predicting device failure one day in advance. Due to the critical nature of early failure detection, **recall on the positive class (failures)** is prioritized over overall accuracy or precision.

> **Why Recall Matters Most:**  
> In predictive maintenance, the cost of a missed failure (false negative) can result in equipment breakdown, service interruption, or safety risks. In contrast, a false alarm (false positive) may only trigger an unnecessary inspection.  
> Therefore, our priority is to **maximize the number of correctly identified failures**, even if it means tolerating some false positives.

> **Recall Formula:**  

Recall = TP / (TP + FN)  


The models were trained on SMOTE-balanced data and evaluated on a realistic, imbalanced test set using a 20% device-based split.

### 📊 Summary Table

| Model                  | Recall (Failures) | True Positives | False Negatives | False Positives | Accuracy |
|------------------------|------------------:|----------------:|-----------------:|-----------------:|---------:|
| Random Forest          | 0.0000            | 0               | 22              | 118              | 0.9945   |
| Balanced RF            | 0.0455            | 1               | 21              | 107              | 0.9950   |
| Logistic Regression    | 0.6818            | 15              | 7               | 1401             | 0.9447   |
| XGBoost                | 0.0455            | 1               | 21              | 71               | 0.9964   |
| HistGradientBoosting   | 0.1818            | 4               | 18              | 138              | 0.9939   |
| LightGBM               | 0.1364            | 3               | 19              | 103              | 0.9952   |

### ✅ Best Performing Model (based on Recall)

The best-performing model for this challenge is **Logistic Regression**, achieving a recall of **0.6818** on the failure class. It correctly identified **15 out of 22 real failures**, far outperforming other models in this critical metric.

While this model generated **a large number of false positives (1,401)**, this is acceptable in the context of **predictive maintenance**, where **missing a failure is costlier than triggering unnecessary alerts**.

### ⚖️ Alternative Balanced Option

The **HistGradientBoostingClassifier** offered a more balanced performance:
- **Recall**: 0.1818 (4 of 22 failures detected)
- **False Positives**: 138 (far lower than Logistic Regression)
- This may be a good choice in environments where **alert fatigue** is a concern.

### 📉 Reducing Alert Fatigue

We can work on reducing alert fatigue by decreasing the number of false positives in the two most promising models. To achieve this, we will apply **decision threshold adjustment**, which allows us to fine-tune the balance between sensitivity (recall) and precision by modifying the cutoff used to classify a prediction as a failure.  
However, increasing the threshold usually comes with a **penalty in true positives**, meaning that while we reduce false alarms, we may also miss some actual failures.

### 📊 Logistic Regression – Threshold Comparison

| Threshold | True Positives (TP) | False Negatives (FN) | False Positives (FP) | Comments |
|-----------|---------------------|-----------------------|-----------------------|----------|
| **0.10**  | 22                  | 0                     | 25,442                | ✅ Captures all failures, but generates many false alarms |
| **0.20**  | 21                  | 1                     | 24,470                | High recall with slightly fewer false positives |
| **0.25**  | 21                  | 1                     | 21,809                | Balanced start — good recall and lower FP |
| **0.30**  | 19                  | 3                     | 16,153                | Reduced FP, but starts missing failures |
| **0.40**  | 15                  | 7                     | 4,428                 | Moderate precision, lower recall |
| **0.50**  | 15                  | 7                     | 1,401                 | ⚖️ Best balance between precision and recall |
| **0.60**  | 14                  | 8                     | 835                   | 📉 Fewer false positives, but misses more failures |

### 🔍 Key Insight

- Lower thresholds (e.g., 0.10) **maximize recall** but result in **very high false positives**, potentially overwhelming response teams.
- Higher thresholds (e.g., 0.60) **reduce false positives**, but **miss more true failures**, which might be risky in critical systems.
- The default threshold of **0.50** offers a **balanced trade-off**, making it a strong candidate for production use, unless the cost of missing a failure is too high.


In [19]:
print("=== ANALYSIS WITH DECISION THRESHOLD ADJUSTMENT (predict_proba) ===")

# --- Models to analyze ---
# We selected the two most promising ones from your initial analysis.
models_to_analyze = {
    "Logistic Regression": models["Logistic Regression"],
    "HistGradientBoosting": models["HistGradientBoosting"]
}

# --- Thresholds to test ---
# You can add or remove values from this list.
# Lower values increase recall; higher values increase precision.
thresholds = [0.1, 0.2, 0.25, 0.3, 0.4, 0.5, 0.6]

# --- Loop to evaluate each model at each threshold ---
for model_name, model in models_to_analyze.items():
    print(f"\n####### Model Analysis: {model_name} #######")
    
    # Get failure probabilities (class 1) once for the model
    y_pred_probs = model.predict_proba(X_test_scaled)[:, 1]
    
    for threshold in thresholds:
        print(f"\n--- Testing with Threshold = {threshold:.2f} ---")
        
        # Apply the threshold to create final predictions
        # If probability >= threshold, predict 1 (failure); otherwise, 0.
        y_pred_custom = (y_pred_probs >= threshold).astype(int)
        
        # Compute confusion matrix
        cm = confusion_matrix(y_test, y_pred_custom)
        tn, fp, fn, tp = cm.ravel()
        
        print(f"Confusion Matrix: \n{cm}")
        print(f"  - True Positives (Failures Detected): {tp}")
        print(f"  - False Negatives (Missed Failures):  {fn}")
        print(f"  - False Positives (False Alarms):     {fp}")

=== ANALYSIS WITH DECISION THRESHOLD ADJUSTMENT (predict_proba) ===

####### Model Analysis: Logistic Regression #######

--- Testing with Threshold = 0.10 ---
Confusion Matrix: 
[[   13 25442]
 [    0    22]]
  - True Positives (Failures Detected): 22
  - False Negatives (Missed Failures):  0
  - False Positives (False Alarms):     25442

--- Testing with Threshold = 0.20 ---
Confusion Matrix: 
[[  985 24470]
 [    1    21]]
  - True Positives (Failures Detected): 21
  - False Negatives (Missed Failures):  1
  - False Positives (False Alarms):     24470

--- Testing with Threshold = 0.25 ---
Confusion Matrix: 
[[ 3646 21809]
 [    1    21]]
  - True Positives (Failures Detected): 21
  - False Negatives (Missed Failures):  1
  - False Positives (False Alarms):     21809

--- Testing with Threshold = 0.30 ---
Confusion Matrix: 
[[ 9302 16153]
 [    3    19]]
  - True Positives (Failures Detected): 19
  - False Negatives (Missed Failures):  3
  - False Positives (False Alarms):     16153

## 📊 Model Performance Comparison – With and Without PCA

PCA was applied to reduce the original 27 features to 6 components (95% of the variance retained). The table below shows the effect on model performance, particularly True Positives (TP), False Negatives (FN), False Positives (FP), and Recall.

| Model                   | PCA? | TP | FN | FP   | Recall   | Comments                                     |
|------------------------|------|----|----|------|----------|----------------------------------------------|
| Logistic Regression    | ❌ No  | 15 |  7 | 1401 | **0.6818** | ✅ Best recall, but many false alarms         |
| Logistic Regression    | ✅ Yes | 14 |  8 | 949  | 0.6364   | ⚖️ Fewer FPs, slight drop in recall          |
| HistGradientBoosting   | ❌ No  | 5  | 17 | 137  | 0.2273   | Low recall, few false positives              |
| HistGradientBoosting   | ✅ Yes | 9  | 13 | 1098 | 0.4091   | 🔼 Significant recall improvement, more FPs   |
| LightGBM               | ❌ No  | 3  | 19 | 103  | 0.1364   | Few false positives, low recall              |
| LightGBM               | ✅ Yes | 10 | 12 | 1214 | 0.4545   | 🔼 Big recall gain, but higher FP count       |
| XGBoost                | ❌ No  | 1  | 21 | 71   | 0.0455   | Very low recall                              |
| XGBoost                | ✅ Yes | 4  | 18 | 649  | 0.1818   | Moderate recall gain, more false positives   |
| Balanced RF            | ❌ No  | 1  | 21 | 107  | 0.0455   | Low recall                                   |
| Balanced RF            | ✅ Yes | 5  | 17 | 319  | 0.2273   | 🔼 Better recall, modest increase in FP       |
| Random Forest          | ❌ No  | 0  | 22 | 118  | 0.0000   | 🚫 No failures detected                       |
| Random Forest          | ✅ Yes | 6  | 16 | 325  | 0.2727   | 🔼 Significant gain in detection              |


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.metrics import confusion_matrix, classification_report
from imblearn.over_sampling import SMOTE
from sklearn.ensemble import RandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier
from sklearn.linear_model import LogisticRegression
from xgboost import XGBClassifier
from sklearn.ensemble import HistGradientBoostingClassifier
from lightgbm import LGBMClassifier

# --- Required Imports ---


# --- Convert and sort the date column ---
df['date'] = pd.to_datetime(df['date'])
df = df.sort_values(['device', 'date']).reset_index(drop=True)

# --- Create the target column ---
df['target'] = df.groupby('device')['failure'].shift(-1).fillna(0)

# --- Rolling statistics per device ---
rolling_window = 5
attr_cols = [f'attribute{i}' for i in range(1, 10)]

for col in attr_cols:
    df[f'{col}_roll_mean'] = df.groupby('device')[col].transform(
        lambda x: x.rolling(window=rolling_window, min_periods=1).mean()
    )
    df[f'{col}_roll_std'] = df.groupby('device')[col].transform(
        lambda x: x.rolling(window=rolling_window, min_periods=1).std().fillna(0)
    )

# --- Stratified split by devices with/without failure ---
device_failures = df.groupby('device')['failure'].max()
failed_devices = device_failures[device_failures == 1].index
healthy_devices = device_failures[device_failures == 0].index

train_failed, test_failed = train_test_split(failed_devices, test_size=0.2, random_state=42)
train_healthy, test_healthy = train_test_split(healthy_devices, test_size=0.2, random_state=42)

train_devices = list(train_failed) + list(train_healthy)
test_devices = list(test_failed) + list(test_healthy)

train_df = df[df['device'].isin(train_devices)].copy()
test_df = df[df['device'].isin(test_devices)].copy()

# --- Features ---
feature_cols = attr_cols + [f'{col}_roll_mean' for col in attr_cols] + [f'{col}_roll_std' for col in attr_cols]

X_train = train_df[feature_cols]
y_train = train_df['target']
X_test = test_df[feature_cols]
y_test = test_df['target']

# --- Normalization ---
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# --- Balancing with SMOTE ---
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X_train_scaled, y_train)

# --- Dimensionality reduction with PCA ---
pca = PCA(n_components=0.95)  # Keep 95% of the variance
X_resampled_pca = pca.fit_transform(X_resampled)
X_test_pca = pca.transform(X_test_scaled)

print(f"Original feature count: {X_resampled.shape[1]}")
print(f"PCA-reduced feature count: {X_resampled_pca.shape[1]}")

# --- Candidate models ---
models = {
    "Random Forest": RandomForestClassifier(random_state=42),
    "Balanced RF": BalancedRandomForestClassifier(random_state=42),
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "XGBoost": XGBClassifier(use_label_encoder=False, eval_metric='logloss'),
    "HistGradientBoosting": HistGradientBoostingClassifier(),
    "LightGBM": LGBMClassifier()
}

def evaluate_model(name, model, X_test, y_test):
    """
    Train and evaluate a given model, printing its performance.
    """
    y_pred = model.predict(X_test)
    print(f"\n=== {name} ===")
    print(confusion_matrix(y_test, y_pred))

# --- Train each model and evaluate using PCA-transformed data ---
for name, model in models.items():
    model.fit(X_resampled_pca, y_resampled)
    evaluate_model(name, model, X_test_pca, y_test)

Original feature count: 27
PCA-reduced feature count: 6

=== Random Forest ===
[[25130   325]
 [   16     6]]

=== Balanced RF ===
[[25136   319]
 [   17     5]]

=== Logistic Regression ===
[[24506   949]
 [    8    14]]


c:\Users\Robson\AppData\Local\Programs\Python\Python313\Lib\site-packages\xgboost\training.py:183: UserWarning: [08:26:30] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)



=== XGBoost ===
[[24806   649]
 [   18     4]]

=== HistGradientBoosting ===
[[24357  1098]
 [   13     9]]
[LightGBM] [Info] Number of positive: 98933, number of negative: 98933
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.003526 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 1530
[LightGBM] [Info] Number of data points in the train set: 197866, number of used features: 6
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000

=== LightGBM ===
[[24241  1214]
 [   12    10]]


c:\Users\Robson\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
